internal/exec/stages/files/files.go

```diff
@@ -79,57 +80,132 @@ func (s stage) Run(config types.Config) bool {
 	return true
 }
 
-// createFilesystemsFiles creates the files described in config.Storage.Filesystems.
-func (s stage) createFilesystemsFiles(config types.Config) error {
+// createFilesystemsEntries creates the files described in config.Storage.{Files,Directories}.
+func (s stage) createFilesystemsEntries(config types.Config) error {
 	if len(config.Storage.Filesystems) == 0 {
 		return nil
 	}
 	s.Logger.PushPrefix("createFilesystemsFiles")
 	defer s.Logger.PopPrefix()
 
-	fileMap, err := s.mapFilesToFilesystems(config)
+	entryMap, err := s.mapEntriesToFilesystems(config)
 	if err != nil {
 		return err
 	}
 
-	for fs, f := range fileMap {
-		if err := s.createFiles(fs, f); err != nil {
+	for fs, f := range entryMap {
+		if err := s.createEntries(fs, f); err != nil {
 			return fmt.Errorf("failed to create files: %v", err)
 		}
 	}
 
 	return nil
 }
 
-// mapFilesToFilesystems builds a map of filesystems to files. If multiple
-// definitions of the same filesystem are present, only the final definition is
-// used.
-func (s stage) mapFilesToFilesystems(config types.Config) (map[types.Filesystem][]types.File, error) {
-	files := map[string][]types.File{}
-	for _, f := range config.Storage.Files {
-		files[f.Filesystem] = append(files[f.Filesystem], f)
+// filesystemEntry represent a thing that knows how to create itself.
+type filesystemEntry interface {
+	create(l *log.Logger, u util.Util) error
+}
+
+type fileEntry types.File
+
+func (tmp fileEntry) create(l *log.Logger, u util.Util) error {
+	f := types.File(tmp)
+	file := util.RenderFile(l, f)
+	if file == nil {
+		return fmt.Errorf("failed to resolve file %q", f.Path)
+	}
+
+	if err := l.LogOp(
+		func() error { return u.WriteFile(file) },
+		"writing file %q", string(f.Path),
+	); err != nil {
+		return fmt.Errorf("failed to create file %q: %v", file.Path, err)
+	}
+
+	return nil
+}
+
+type dirEntry types.Directory
+
+func (tmp dirEntry) create(l *log.Logger, u util.Util) error {
+	d := types.Directory(tmp)
+	err := l.LogOp(func() error {
+		path := u.JoinPath(string(d.Path))
+		if err := util.MkdirForFile(path); err != nil {
+			return err
+		}
+
+		if err := os.Mkdir(path, os.FileMode(d.Mode)); err != nil {
+			return err
+		}
+
+		// Set the ownership and mode are as requested (since Mkdir doesn't set ownership and can be affect by the sticky bit)
```

Comment:
The mkdir syscall will apply umask to the mode, chmod is explicit.  I was likely being conservative when I wrote that code originally, we could just take care to clear the umask of the executing environment for ignition, but manual testing would then potentially produce different results if the umask was something else.

Setting the mode explicitly takes umask out of the equation...